File: 0918.环形子数组的最大和/0918-环形子数组的最大和.c
int maxSubarraySumCircular(int* nums, int numsSize){
    int dp = nums[0],max = dp, sum = dp,min = 0;
    for(int i = 1; i < numsSize; i++){
        sum += nums[i];
        dp = nums[i] + fmax(dp, 0);
        max = fmax(dp, max);
    }
    dp = nums[0];
    for(int i = 1; i < numsSize -1; i++){
        dp = nums[i] + fmin(0,dp);
        min = fmin(dp,min);
    }
    return fmax(sum-min,max);
}

File: 0918.环形子数组的最大和/0918-环形子数组的最大和.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int dp = nums[0],max = dp, sum = dp,min = 0;
        for(int i = 1; i < nums.length; i++){
            sum += nums[i];
            dp = nums[i] + Math.max(dp, 0);
            max = Math.max(dp, max);
        }
        dp = nums[0];
        for(int i = 1; i < nums.length -1; i++){
            dp = nums[i] + Math.min(0,dp);
            min = Math.min(dp,min);
        }
        return Math.max(sum-min,max);
    }
}